File: apps/fsharp/src/MachineLearning/Chapter13/Main.fs
module MachineLearning.Chapter13.Main

open System.IO
open MachineLearning.Dataset
open MachineLearning.Chapter13.BostonStandardized
open MachineLearning.Chapter13.Pca
open MachineLearning.Chapter13.StatsPca

[<Literal>]
let Threshold = 0.8

[<Literal>]
let TopK = 3

[<Literal>]
let ComponentsToExplain = 2

/// FSharp.Stats の結果と一致したとみなす差の上限
[<Literal>]
let Tolerance = 1e-9

let private within (a: float) (b: float) : bool = abs (a - b) <= Tolerance

/// 2 つの配列を先頭から組にして、agree が真になる組の数を数える
let private countAgreed (agree: 'T -> 'T -> bool) (a: 'T[]) (b: 'T[]) : int =
    Array.map2 agree a b |> Array.filter id |> Array.length

let private formatLoadings (loadings: (string * float) list) : string =
    loadings
    |> List.map (fun (column, value) -> $"{column} {value:F3}")
    |> String.concat ", "

/// Boston.csv を標準化して主成分分析し、寄与率と主成分への影響が大きい列、FSharp.Stats との一致数を表示する
let run (print: string -> unit) : unit =
    let table = loadStandardizedBoston (Path.Combine(dataDir (), "Boston.csv"))
    let model = fitPca table.Columns.Length table.X
    let ratios = model.ExplainedVarianceRatio
    let needed = componentsNeeded Threshold ratios
    let shown = ratios |> Array.truncate needed
    print $"データ件数: {table.X.Length}, 列数: {table.Columns.Length}"

    shown
    |> Array.mapi (fun i ratio -> $"PC{i + 1} {ratio:F4}")
    |> String.concat ", "
    |> fun text -> print $"寄与率: {text}"

    print $"累積寄与率が {Threshold} に届く主成分の数: {needed}（累積寄与率 {Array.sum shown:F4}）"

    for i in 0 .. ComponentsToExplain - 1 do
        let loadings = topLoadings TopK table.Columns model.Components[i]
        print $"第 {i + 1} 主成分で影響の大きい列: {formatLoadings loadings}"

    let library = fitPcaWithFSharpStats table.Columns.Length table.X
    let ratiosAgreed = countAgreed within ratios library.ExplainedVarianceRatio

    let componentsAgreed =
        countAgreed (Array.forall2 within) model.Components library.Components

    print
        $"FSharp.Stats の PCA と一致した数（許容誤差 {Tolerance}）: 寄与率 {ratiosAgreed}/{ratios.Length}, 主成分 {componentsAgreed}/{model.Components.Length}"

File: apps/scala/src/main/scala/machinelearning/chapter13/Main.scala
package machinelearning.chapter13

import java.nio.file.Paths
import java.util.Locale
import machinelearning.dataset.DataDir

/** ボストンの住宅価格の 15 列を主成分分析し、寄与率と主成分の意味を表示する。 */
object Main:
  private val Threshold = 0.8
  private val TopK = 3
  private val ComponentsToExplain = 2

  def run(print: String => Unit): Unit =
    val features = BostonPca.load(Paths.get(DataDir.current(), "Boston.csv"))
    val columns = features.head.columns
    val x = BostonPca.toMatrix(features)
    val model = Pca.fit(x, columns.size)
    val ratios = model.explainedVarianceRatio
    val needed = Pca.componentsNeeded(ratios, Threshold)
    val cumulative = ratios.take(needed).sum

    print(s"データ件数: ${x.rowCount}, 列数: ${columns.size}")
    print(
      "寄与率: " + ratios
        .take(needed)
        .zipWithIndex
        .map((ratio, i) => s"PC${i + 1} ${format(ratio, 4)}")
        .mkString(", ")
    )
    print(s"累積寄与率が $Threshold に届く主成分の数: $needed（累積寄与率 ${format(cumulative, 4)}）")
    model.components.rows.take(ComponentsToExplain).zipWithIndex.foreach { (component, i) =>
      val loadings = Pca
        .topLoadings(component, columns, TopK)
        .map(loading => s"${loading.column} ${format(loading.value, 3)}")
        .mkString(", ")
      print(s"第 ${i + 1} 主成分で影響の大きい列: $loadings")
    }

  private def format(value: Double, digits: Int): String =
    String.format(Locale.ROOT, s"%.${digits}f", value)
